packages/superset-ui-preset-chart-xy/src/encodeable/parsers/extractScale.ts

```diff
@@ -1,38 +1,246 @@
-import { CategoricalColorNamespace } from '@superset-ui/color';
-import { scaleOrdinal } from 'd3-scale';
+import { CategoricalColorNamespace, CategoricalColorScale } from '@superset-ui/color';
+import {
+  ScaleOrdinal,
+  ScaleLinear,
+  ScaleLogarithmic,
+  ScalePower,
+  ScaleTime,
+  ScaleQuantile,
+  ScaleQuantize,
+  ScaleThreshold,
+  ScalePoint,
+  ScaleBand,
+  scaleLinear,
+  scaleLog,
+  scalePow,
+  scaleSqrt,
+  scaleTime,
+  scaleUtc,
+  scaleQuantile,
+  scaleQuantize,
+  scaleThreshold,
+  scaleOrdinal,
+  scalePoint,
+  scaleBand,
+} from 'd3-scale';
 import { Value } from 'vega-lite/build/src/fielddef';
-import isEnabled from '../utils/isEnabled';
-import { isScaleFieldDef, ChannelDef, isPositionFieldDef } from '../types/FieldDef';
+import { Type } from 'vega-lite/build/src/type';
+import { ScaleType } from 'vega-lite/build/src/scale';
+import { isNonValueDef, ChannelDef } from '../types/FieldDef';
+import isDisabled from '../utils/isDisabled';
+import { ChannelType } from '../types/Channel';
+import { Scale } from '../types/Scale';
 
-export default function extractScale<Output extends Value = Value>(
+export interface ScaleAgent<Output extends Value> {
+  config: Scale<Output>;
+  setDomain: (newDomain: number[] | string[] | boolean[] | Date[]) => void;
+  encodeValue: (value: number | string | boolean | null | undefined | Date) => Output;
+  scale:
+    | CategoricalColorScale
+    | ScaleLinear<Output, Output>
+    | ScaleLogarithmic<Output, Output>
+    | ScalePower<Output, Output>
+    | ScaleLogarithmic<Output, Output>
+    | ScaleTime<Output, Output>
+    | ScaleQuantile<Output>
+    | ScaleQuantize<Output>
+    | ScaleThreshold<number | string | Date, Output>
+    | ScaleOrdinal<{ toString(): string }, Output>
+    | ScalePoint<{ toString(): string }>
+    | ScaleBand<{ toString(): string }>;
+}
+
+export interface ScaleTypeToD3ScaleType<Output> {
+  [ScaleType.LINEAR]: ScaleLinear<Output, Output>;
+  [ScaleType.LOG]: ScaleLogarithmic<Output, Output>;
+  [ScaleType.POW]: ScalePower<Output, Output>;
+  [ScaleType.SQRT]: ScalePower<Output, Output>;
+  [ScaleType.SYMLOG]: ScaleLogarithmic<Output, Output>;
```

Comment:
wow, you really considered all the possibilities ! 👍

Reply:
![image](https://user-images.githubusercontent.com/1659771/55196933-304b3b80-516e-11e9-9751-16aaa628cf4b.png)